### ISSUE TITLE
Stale connections pinging the MRE, causing a deadlock

### ISSUE THREAD

**Issue by willneedit** [Microsoft/mixed-reality-extension-sdk#104]
On running the Stargate in a larger scale, I noticed that I got connection attempts from two distinct IP addresses which spawned the MRE's in an otherwise empty space.

Biggest problem is that since these 'stale' instance become authoritative for the geometry and animations displayed, they prevent the MRE from functioning for every other user at all.

Resetting the space might get rid of it, and (untested) blocking the IP's at link level might mitigate the problem.

Additionally, Heroku cuts idle connections after 55 seconds, but the logs show these IP addresses over and over again, so they definitely reconnect.

**Comment by eanders-MS** [comment]
Help me understand. For the MRE connection to be attempted, the hosting space has to be running on at least one client. What makes the MRE connection 'stale'?

**Comment by willneedit** [comment]
I don't know the whole mechanics, but I think those two connection attempts come from users who once visited this space and either idled out/left in a nonstandard way/whatever. So the space itself registers as 0/30 users, but the clients didn't get the notice, so they do try to contact the MRE's of the space they *think* they still belong to.

Logs look much like the following:

```
2019-01-10T09:32:30.905343+00:00 app[web.1]: Unregistering gate for ID ttttttt
2019-01-10T09:32:30.906398+00:00 heroku[router]: at=error code=H15 desc="Idle connection" method=GET path="/app?name,sgdialcomp&id,ttttttt" host=willneedit-mre.herokuapp.com request_id=1dd3ddf7-fc11-4b32-b1b6-6b5374a0db06 fwd="IP REDACTED" dyno=web.1 connect=0ms service=55011ms status=503 bytes= protocol=http
2019-01-10T09:32:32.454617+00:00 app[web.1]: Starting: stargate in D/Hn2lIRJ889sdf2 ...
2019-01-10T09:32:32.454777+00:00 app[web.1]: Registering gate for ID ttttttt
```

The 'Unregistering' line is emitted in onStop(), and the 'Registering' line is when the first user is detected. Nothing unusual in itself, but it happens constantly whenever Heroku cuts the connection as 'idle', forcing some client to reconnect... and on a world that is listed as 0/30 users present.

**Comment by willneedit** [comment]
UPDATE: With thosee two IP addresses still going at it even after a full day, now I'm less inclined to believe it's a misbehaving client but something more malicious. I've blacklisted these two IPs successfully, making Cheyenne mountain operational again.

**Comment by willneedit** [comment]
Hmmm. Explicitely added a log line in OnUserJoined right now, but it doesn't show up. So no user announcement.

**Comment by eanders-MS** [comment]
A look at your application's network traffic might be interesting. It would tell us how far the user is getting through the connection process. You can enable verbose network logging by putting these lines near the top of your server.ts:
```ts
import { log } from '@microsoft/mixed-reality-extension-sdk/built/log';

log.enable('network');
```

**Comment by willneedit** [comment]
By now everything seems to have calmed down on that front, but I'll definitely be adding network logs to the output, in case it could or when happen again - or the moment it happens again.

I'll keep an eye open.

**Comment by eanders-MS** [comment]
Oh that's interesting. The log above shows the handshake between the multipeer adapter and your app. There is no traffic to or from the client. Somehow they're getting by the protocol check, but then for some reason nothing additional is sent (by the client). It seems we have two problems here:
- This is the first client connection so we mark it as the authoritative peer, but we're doing this too early, before they're fully joined. This makes your app effectively broken until that connection goes away. No other users can join.
- We're starting up your app too early. We should do this after we have a fully joined client, not before.

This is really valuable, thank you. My focus next week will be making this process more robust.

**Comment by willneedit** [comment]
In addition - since the 'Gate has to use an enclosure and it maintains its own WebSocket connection, I could see these connection attempts from this WebSocket dropping in - so the enclosure itself must be loaded and active on the affected client as well. Which means it looks like a situation where the Altspace servers and the client strongly disagree on where the client is supposed to be ... or online at all.

Just taking a guess. Android platform user simply taking off his headset, making the client go into a semi-dormant state which causes Altspace to disconnect the user from the service?

**Comment by eanders-MS** [comment]
Fixed by #155
